documentation/technical_writing_slides.md

```diff
@@ -0,0 +1,452 @@
+
+---
+
+## Error Messages (2/5)
+
+Examples from preCICE before Spring 2020:
+
+- *Safety Factor must be positive or 0*
+- *Data with name "Forces" is not defined on mesh with ID 1.*
+- *At least two participants need to be defined!*
+
+What is wrong with these?
+
+---
+
+## Error Messages (3/5)
+
+Examples from preCICE after Spring 2020 (error messages sprint):
+
+- **Give context**:
+    - *Participant "Fluid" uses mesh "FluidMesh" with safety-factor="-0.5". Please use a positive or zero safety-factor instead.*
+- **Give advice**:
+    - *Data with name "Forces" is not defined on mesh "FluidMesh". Please add `<use-data name="Forces"/>` under `<mesh name="FluidMesh/>`.*
+- **Improve tone**:
+    - *In the preCICE configuration, only one participant is defined. One participant makes no coupled simulation. Please add at least another one.*
+
+---
+
+## Error Messages (4/5)
+
+Tips from [Write the Docs](https://www.writethedocs.org/guide/writing/style-guides/#error-messages):
+
+- Provide explicit indication that something has gone wrong.
+- Write like a human, not a robot.
+- Don’t blame the user – be humble.
+- Make the message short and meaningful.
+- Include precise descriptions of exact problems.
+- Offer constructive advice on how to fix the problem.
+
+---
+
+## Error Messages (5/5)
+
+Who did something wrong? The software or the user?
+
+- The user -> error
+- The software -> assertion
+- Make this transparent for user
+
+---
+
+## Commit Messages (1/3)
+
+Same story again:
+
+- Consistent
+- Descriptive and concise (such that complete history becomes skimmable)
+- Explain the "why" (the "what" is covered in the diff)
+
+---
+
+## Commit Messages (2/3)
+
+[The seven rules of a great Git commit message](https://chris.beams.io/git-commit/):
+
+- Separate subject from body with a blank line.
+- Limit the subject line to 50 characters.
+- Capitalize the subject line.
+- Do not end the subject line with a period.
+- Use the imperative mood in the subject line.
+- Wrap the body at 72 characters.
+- Use the body to explain what and why vs. how.
+
+---
+
+## Commit Messages (3/3)
+
+Tags can also be useful (e.g. `[Bugfix]`), but consistency is key. Our [lecture material](https://github.com/Simulation-Software-Engineering/Lecture-Material/commits/main) is a bad example. :grin:
+
+---
+
+## API Documentation
+
+- All basics apply (be consistent, descriptive, and concise; explain the why, ...)
+- Consistency is also great for automatization: rendering, interlinking, suggestions by IDE, ... (next week)
+- Different syntax (e.g. [Google style for Python](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings), [numpydoc style](https://numpydoc.readthedocs.io/en/latest/), [doxygen for C++](https://www.doxygen.nl), [Javadoc](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html))
```

Comment:
Doxygen for almost everything :smile: 

> ... C++ sources, but it also supports other popular programming languages such as C, Objective-C, C#, PHP, Java, Python, IDL (Corba, Microsoft, and UNO/OpenOffice flavors), Fortran, VHDL and to some extent D.

I contributed to some Matlab code that used Doxygen.